### PR TITLE
Fix process response getting duplicated requests

### LINKED ISSUE #76: Version v1.6.0 is much unfriendly towards non-HTTP requests
## 1. Summary of the Issue

The middleware may attempt to retry non-HTTP requests since version v1.6.0, or precisely, commit 11879301.
And such retrying mechanism is not properly implemented:
- If the original request comes with `dont_filter=False`, it'll be dropped when being enqueued the 2nd time in the scheduler for being duplicate.
- If the original request comes with `dont_filter=True`, `CrawleraMiddleware` would attempt to retry it indefinitely.

## 2. To Reproduce the Issue

Sample spider for reproducing the issue:

```python
# coding: utf8

import scrapy


class TestSpider(scrapy.Spider):
    name = 'test'
    custom_settings = {
        'DOWNLOADER_MIDDLEWARES': {
            'scrapy_crawlera.CrawleraMiddleware': 610,
        },
        'CRAWLERA_ENABLED': True,
        'CRAWLERA_APIKEY': 'foobar',  # no worries. this'll work
    }

    def start_requests(self):
        yield scrapy.Request('file:///tmp/foobar.txt')

    def parse(self, response):
        self.logger.info(response.text)
```
Sample logs with `scrapy-crawlera==1.6.0`:
```
$ echo -n baz > /tmp/foobar.txt && scrapy runspider test.py
2019-06-14 04:05:38 [scrapy.utils.log] INFO: Scrapy 1.6.0 started (bot: scrapybot)
2019-06-14 04:05:38 [scrapy.utils.log] INFO: Versions: lxml 4.3.2.0, libxml2 2.9.9, cssselect 1.0.3, parsel 1.5.0, w3lib 1.20.0, Twisted 18.9.0, Python 3.7.3 (default, Mar 26 2019, 21:43:19) - [GCC 8.2.1 20181127], pyOpenSSL 19.0.0 (OpenSSL 1.1.1b  26 Feb 2019), cryptography 2.6.1, Platform Linux-5.1.4-arch1-1-ARCH-x86_64-with-arch
2019-06-14 04:05:38 [scrapy.crawler] INFO: Overridden settings: {'EDITOR': 'vim', 'SPIDER_LOADER_WARN_ONLY': True}
2019-06-14 04:05:38 [scrapy.extensions.telnet] INFO: Telnet Password: 5dd83baa8682f121
2019-06-14 04:05:38 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.memusage.MemoryUsage',
 'scrapy.extensions.logstats.LogStats']
2019-06-14 04:05:38 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy_crawlera.CrawleraMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2019-06-14 04:05:38 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2019-06-14 04:05:38 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2019-06-14 04:05:38 [scrapy.core.engine] INFO: Spider opened
2019-06-14 04:05:38 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2019-06-14 04:05:38 [root] INFO: Using crawlera at http://proxy.crawlera.com:8010 (apikey: foobar)
2019-06-14 04:05:38 [root] INFO: CrawleraMiddleware: disabling download delays on Scrapy side to optimize delays introduced by Crawlera. To avoid this behaviour you can use the CRAWLERA_PRESERVE_DELAY setting but keep in mind that this may slow down the crawl significantly
2019-06-14 04:05:38 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2019-06-14 04:05:38 [scrapy.dupefilters] DEBUG: Filtered duplicate request: <GET file:///tmp/foobar.txt> - no more duplicates will be shown (see DUPEFILTER_DEBUG to show all duplicates)
2019-06-14 04:05:38 [scrapy.core.engine] INFO: Closing spider (finished)
2019-06-14 04:05:38 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{'crawlera/request': 1,
 'crawlera/request/method/GET': 1,
 'downloader/request_bytes': 257,
 'downloader/request_count': 1,
 'downloader/request_method_count/GET': 1,
 'downloader/response_bytes': 22,
 'downloader/response_count': 1,
 'downloader/response_status_count/200': 1,
 'dupefilter/filtered': 1,
 'finish_reason': 'finished',
 'finish_time': datetime.datetime(2019, 6, 14, 3, 5, 38, 235584),
 'log_count/DEBUG': 1,
 'log_count/INFO': 11,
 'memusage/max': 54403072,
 'memusage/startup': 54403072,
 'scheduler/dequeued': 1,
 'scheduler/dequeued/memory': 1,
 'scheduler/enqueued': 1,
 'scheduler/enqueued/memory': 1,
 'start_time': datetime.datetime(2019, 6, 14, 3, 5, 38, 229586)}
2019-06-14 04:05:38 [scrapy.core.engine] INFO: Spider closed (finished)
```
Sample logs with `scrapy-crawlera==1.5.1`:
```
$ echo -n baz > /tmp/foobar.txt && scrapy runspider test.py
2019-06-14 04:04:58 [scrapy.utils.log] INFO: Scrapy 1.6.0 started (bot: scrapybot)
2019-06-14 04:04:58 [scrapy.utils.log] INFO: Versions: lxml 4.3.2.0, libxml2 2.9.9, cssselect 1.0.3, parsel 1.5.0, w3lib 1.20.0, Twisted 18.9.0, Python 3.7.3 (default, Mar 26 2019, 21:43:19) - [GCC 8.2.1 20181127], pyOpenSSL 19.0.0 (OpenSSL 1.1.1b  26 Feb 2019), cryptography 2.6.1, Platform Linux-5.1.4-arch1-1-ARCH-x86_64-with-arch
2019-06-14 04:04:58 [scrapy.crawler] INFO: Overridden settings: {'EDITOR': 'vim', 'SPIDER_LOADER_WARN_ONLY': True}
2019-06-14 04:04:58 [scrapy.extensions.telnet] INFO: Telnet Password: 930d9b1988e0e136
2019-06-14 04:04:58 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.memusage.MemoryUsage',
 'scrapy.extensions.logstats.LogStats']
2019-06-14 04:04:58 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy_crawlera.CrawleraMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2019-06-14 04:04:58 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2019-06-14 04:04:58 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2019-06-14 04:04:58 [scrapy.core.engine] INFO: Spider opened
2019-06-14 04:04:58 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2019-06-14 04:04:58 [root] INFO: Using crawlera at http://proxy.crawlera.com:8010 (apikey: foobar)
2019-06-14 04:04:58 [root] INFO: CrawleraMiddleware: disabling download delays on Scrapy side to optimize delays introduced by Crawlera. To avoid this behaviour you can use the CRAWLERA_PRESERVE_DELAY setting but keep in mind that this may slow down the crawl significantly
2019-06-14 04:04:58 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2019-06-14 04:04:58 [scrapy.core.engine] DEBUG: Crawled (200) <GET file:///tmp/foobar.txt> (referer: None)
2019-06-14 04:04:58 [test] INFO: baz
2019-06-14 04:04:58 [scrapy.core.engine] INFO: Closing spider (finished)
2019-06-14 04:04:58 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{'crawlera/request': 1,
 'crawlera/request/method/GET': 1,
 'crawlera/response': 1,
 'crawlera/response/status/200': 1,
 'downloader/request_bytes': 257,
 'downloader/request_count': 1,
 'downloader/request_method_count/GET': 1,
 'downloader/response_bytes': 22,
 'downloader/response_count': 1,
 'downloader/response_status_count/200': 1,
 'finish_reason': 'finished',
 'finish_time': datetime.datetime(2019, 6, 14, 3, 4, 58, 793630),
 'log_count/DEBUG': 1,
 'log_count/INFO': 12,
 'memusage/max': 54611968,
 'memusage/startup': 54611968,
 'response_received_count': 1,
 'scheduler/dequeued': 1,
 'scheduler/dequeued/memory': 1,
 'scheduler/enqueued': 1,
 'scheduler/enqueued/memory': 1,
 'start_time': datetime.datetime(2019, 6, 14, 3, 4, 58, 681959)}
2019-06-14 04:04:58 [scrapy.core.engine] INFO: Spider closed (finished)
```
The same issue may be reproduced in various common environments (different versions of Scrapy, Python, OS, etc.).

## 3. The Cause

This issue is believed to have been caused by improperly designed response handling log. Related code: https://github.com/scrapy-plugins/scrapy-crawlera/blob/v1.6.0/scrapy_crawlera/middleware.py#L178

## 4. Proposed Changes

4.1. Return proper boolean values for non-HTTP requests / responses in methods including [`CrawleraMiddleware._should_enable_for_response`](https://github.com/scrapy-plugins/scrapy-crawlera/blob/v1.6.0/scrapy_crawlera/middleware.py#L254) and [`CrawleraMiddleware._is_enabled_for_request`](https://github.com/scrapy-plugins/scrapy-crawlera/blob/v1.6.0/scrapy_crawlera/middleware.py#L257) by checking the URL scheme.
4.2. Other than retrying requests unconditionally on [this line](https://github.com/scrapy-plugins/scrapy-crawlera/blob/v1.6.0/scrapy_crawlera/middleware.py#L179), do it with a maximum retry limit like in the built-in `RetryMiddleware`, or just make use of that middleware directly. Also, `dont_filter=True` may be assigned if a request needs to be retried here.

### PR DESCRIPTION
**Pull request by denisgermano**
Closes
- fixes #76 

This proposed PR changes to return the actual `response` when is not a response from crawlera, fixing the duplicated request (or indefinitely if `dont_filter=True`)

### CODE REVIEW

**Comment by hcoura** [comment]
Hi @denisgermano thanks for picking it up, overall it looks good and seems like it is indeed fixing my mistakes, the only thing I would like to see are a few tests to cover exactly what @starrify reported on the issue, just so that we reduce the chance of regression.

File: scrapy_crawlera/middleware.py [hunk 1]
```diff
@@ -30,6 +30,7 @@ class CrawleraMiddleware(object):
     force_enable_on_http_codes = []
     max_auth_retry_times = 10
     enabled_for_domain = {}
+    apikey = ""
```
Comment (Gallaecio) [inline scrapy_crawlera/middleware.py]: Is this change necessary or related to the issue?

Comment (denisgermano) [reply scrapy_crawlera/middleware.py]: It is not necessary, just added for the pylint doesn't show a warning of `Instance of 'CrawleraMiddleware' has no 'apikey' member pylint(no-member)` and be explicit that starts as an empty string as default
